From 5fdeff3524a67ee4e2cfa6c8b2c5994bf8a477ea Mon Sep 17 00:00:00 2001
From: Don Darling <don.osc2@gmail.com>
Date: Tue, 16 Mar 2010 15:16:00 -0500
Subject: [PATCH 12/14] v4l2src: update gst_v4l2_get_norm to handle DM6467T and DM365 properly

DM6467T and DM365 use the QUERYSTD ioctl to detect the standard in use by the
connected component.
---
 sys/v4l2/v4l2_calls.c |   35 +++++++++++++++++++++++++++++++++++
 1 files changed, 35 insertions(+), 0 deletions(-)

diff --git a/sys/v4l2/v4l2_calls.c b/sys/v4l2/v4l2_calls.c
index ee52f21..da9c215 100644
--- a/sys/v4l2/v4l2_calls.c
+++ b/sys/v4l2/v4l2_calls.c
@@ -619,8 +619,43 @@ gst_v4l2_get_norm (GstV4l2Object * v4l2object, v4l2_std_id * norm)
   if (!GST_V4L2_IS_OPEN (v4l2object))
     return FALSE;
 
+  #if defined(Platform_dm6467t) || defined(Platform_dm365)
+  /* It might take a few tries to detect the signal -- try up to 300 times */
+  {
+    int retries = 300;
+
+    while (retries-- > 0) {
+
+      /* Break-out of the retry loop on success, and abort if errno is not
+       * EAGAIN.
+       */
+      if (v4l2_ioctl (v4l2object->video_fd, VIDIOC_QUERYSTD, norm) >= 0) {
+        break;
+      }
+      else if (errno != EAGAIN) {
+        goto std_failed;
+      }
+
+      /* Sleep for a microsecond before retrying */
+      usleep(1);
+    }
+
+    /* Abort if all 300 retry attempts failed */
+    if (retries == 0) {
+      goto std_failed;
+    }
+
+    /* On DM6467T, the convention is the set the norm after querying it,
+     * otherwise the driver won't initialize properly.
+     */
+    if (!gst_v4l2_set_norm(v4l2object, *norm)) {
+      goto std_failed;
+    }
+  }
+  #else
   if (v4l2_ioctl (v4l2object->video_fd, VIDIOC_G_STD, norm) < 0)
     goto std_failed;
+  #endif
 
   return TRUE;
 
-- 
1.6.3.3

